core/tracker/tracker.go

```diff
@@ -0,0 +1,252 @@
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of  MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package tracker
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/obolnetwork/charon/core"
+)
+
+//go:generate stringer -type=component
+
+// component refers to a core workflow component.
+type component int
+
+// Core components arranged in the order data flows through them.
+const (
+	scheduler component = iota
+	fetcher
+	consensus
+	validatorAPI
+	parSigDBInternal
+	parSigEx
+	parSigDBThreshold
+	sigAgg
+)
+
+// event represents an event emitted by a core workflow component.
+type event struct {
+	duty      core.Duty
+	component component
+	pubkey    core.PubKey
+}
+
+// Tracker represents the component that listens to events from core workflow components.
+// It identifies where a duty gets stuck in the course of its execution.
+type Tracker struct {
+	input chan event
+
+	// events stores all the events in a particular duty.
+	events    map[core.Duty][]event
+	deadliner core.Deadliner
+	quit      chan struct{}
+}
+
+// NewTracker returns a new Tracker.
+func NewTracker(deadliner core.Deadliner) *Tracker {
+	t := &Tracker{
+		input:     make(chan event),
+		events:    make(map[core.Duty][]event),
+		quit:      make(chan struct{}),
+		deadliner: deadliner,
+	}
+
+	return t
+}
+
+// Run blocks and registers events from each component in tracker's input channel.
+func (t *Tracker) Run(ctx context.Context) error {
+	defer close(t.quit)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case e := <-t.input:
+			// TODO(dhruv): do not store events for expired duties
+			t.events[e.duty] = append(t.events[e.duty], e)
+			t.deadliner.Add(e.duty)
+		case duty := <-t.deadliner.C():
+			failed, failedComponent, failedMsg := analyseFailedDuty(duty, t.events[duty])
+			if failed {
+				t.reportFailedDuties(duty, failedComponent, failedMsg)
+			}
+			// TODO(dhruv): Case of cluster participation (duty success)
+			// t.analyseClusterParticipation()
+			delete(t.events, duty)
+		}
+	}
+}
+
+// analyseFailedDuty analyzes the events for a given duty after the duty's deadline is exceeded.
+func analyseFailedDuty(duty core.Duty, events []event) (bool, component, string) {
+	// Sort in reverse order (see order above).
+	sort.Slice(events, func(i, j int) bool {
+		return events[i].component > events[j].component
+	})
+
+	if events[0].component == sigAgg {
+		// Duty completed successfully
+		return false, sigAgg, ""
+	}
+
+	return true, events[0].component + 1, fmt.Sprintf("%s is failed in %s component", duty.String(), (events[0].component + 1).String())
```

Comment:
I think we can improve the error message. Be more specific. If it got stuck during validatorAPI, we can say "the VC didn't submit a signed duty", if it got stuck at parsigdb then "peers didn't share any/sufficient partial signatures", etc.

Reply:
Agreed.

We can have specific analysis in a separate PR.